idseq_dag/steps/generate_coverage_stats.py

```diff
@@ -62,47 +64,71 @@ def run(self):
                 output_str = ','.join(str(e) for e in output_row)
                 csc.write(output_str + "\n")
 
+    @staticmethod
+    def _process_contig_names(bam_file, contig_names):
+        contig2coverage = {}
+        if len(contig_names) > 0:
+            with log.log_context(f"calc_contig2coverage_chunk", {"bam_file": bam_file, "from": contig_names[0], "to": contig_names[-1]}):
+                with pysam.AlignmentFile(bam_file, "rb") as f:
+                    for c in contig_names:
+                        coverage = []
+                        for pileup_column in f.pileup(contig=c):
+                            coverage.append(pileup_column.get_num_aligned())
+                        sorted_coverage = sorted(coverage)
+                        contig_len = len(coverage)
+                        if contig_len <= 0:
+                            continue
+
+                        avg = sum(coverage)/contig_len
+                        avg2xcnt = 0
+                        cnt0 = 0
+                        cnt1 = 0
+                        cnt2 = 0
+                        for t in coverage:
+                            if t > 2 * avg:
+                                avg2xcnt += 1
+                            if t == 0:
+                                cnt0 += 1
+                            elif t == 1:
+                                cnt1 += 1
+                            elif t == 2:
+                                cnt2 += 1
+
+                        contig2coverage[c] = {
+                            "coverage": coverage,
+                            "avg": avg,
+                            "p0": sorted_coverage[0],
+                            "p100": sorted_coverage[-1],
+                            "p25": sorted_coverage[int(0.25*contig_len)],
+                            "p50": sorted_coverage[int(0.5*contig_len)],
+                            "p75": sorted_coverage[int(0.75*contig_len)],
+                            "avg2xcnt": avg2xcnt / contig_len,
+                            "cnt0": cnt0 / contig_len,
+                            "cnt1": cnt1 / contig_len,
+                            "cnt2": cnt2 / contig_len
+                        }
+        return contig2coverage
 
     @staticmethod
     def calc_contig2coverage(bam_file):
-        CHUNK_SIZE = 100
-        contig2coverage = {}
+        CHUNK_SIZE = 200
```

Comment:
Any reasoning for this particular value?

Reply:
No particular reason for the number 200, just because it is 2X the previous value.

It is worth pointing out that I also arbitrarily established the previous value of 100 when I first created the chunking mechanism for this operation (the original motivation: post processing was being killed by pipeline watchdog for these large samples because they took too long to report any progress, so I needed to chunk this operation to introduce some progress log entries).

However, I noticed each chunk usually take just a few seconds, and since this parallel operation opens and closes the alignment file for each chunk (to avoid multiprocessing issues) I tried to reduce the number of times that will happen by doubling the size of the chunk.